JustSaying.Messaging/MessageProcessingStrategies/Throttled.cs

```diff
@@ -1,109 +1,75 @@
 ﻿using System;
-using System.Collections.Generic;
-using System.Linq;
+using System.Diagnostics;
+using System.Threading;
 using System.Threading.Tasks;
 using JustSaying.Messaging.Monitoring;
 
 namespace JustSaying.Messaging.MessageProcessingStrategies
 {
     public class Throttled : IMessageProcessingStrategy
     {
-        private readonly Func<int> _maxWorkersProducer;
-
         private readonly IMessageMonitor _messageMonitor;
-        private readonly List<Task> _activeTasks;
+        private readonly SemaphoreSlim _semaphore;
 
         public Throttled(int maxWorkers, IMessageMonitor messageMonitor)
-            : this(() => maxWorkers, messageMonitor)
-        {}
-
-        public Throttled(Func<int> maxWorkersProducer,
-            IMessageMonitor messageMonitor)
         {
-            _maxWorkersProducer = maxWorkersProducer;
             _messageMonitor = messageMonitor;
-
-            _activeTasks = new List<Task>(_maxWorkersProducer());
+            MaxWorkers = maxWorkers;
+            _semaphore = new SemaphoreSlim(maxWorkers, maxWorkers);
         }
 
-        public int MaxWorkers
+        public void StartWorker(Func<Task> action)
         {
-            get { return _maxWorkersProducer(); }
+            var messageProcessingTask = new Task<Task>(() => ReleaseOnCompleted(action));
+            _semaphore.Wait();

```

Comment:
We could, but if we got here and ```_semaphore.Wait()``` blocks then we have entered a very edge case race condition, and I think we should block the loop momentarily which is what should have happened back when we wait for workers (but multiple threads will get released a the same time, this is the problem)